examples/move/scratch_off/sources/game.move

```diff
@@ -0,0 +1,202 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+module scratch_off::game {
+
+    use std::vector;
+    use sui::coin::{Self, Coin};
+    use sui::balance::{Self, Balance};
+    use sui::object::{Self, UID, ID};
+    use sui::tx_context::{Self, TxContext};
+    use sui::dynamic_object_field as dof;
+    use sui::transfer;
+    use sui::event;
+    use sui::bls12381::bls12381_min_pk_verify;
+    use sui::hash::blake2b256;
+    use scratch_off::math;
+
+    // --------------- Constants ---------------
+    const DEFAULT_TICKET_NUMBER: u64 = 100_000_000;
+
+    const EInvalidInputs: u64 = 0;
+    const EInvalidBlsSig: u64 = 1;
+
+    // --------------- Events ---------------
+    struct NewDrawing<phantom T> has copy, drop {
+        ticket_id: ID,
+        player: address,
+    }
+
+    struct Ticket has key, store {
+		id: UID,
+        /// Which store this ticket belongs to
+        convenience_store_id: ID,
+        /// Original receiver address, but on evaluation this is set to the ctx::sender addr
+        player: address,
+	}
+    
+	struct ConvenienceStore<phantom Asset> has key {
+        id: UID,
+        creator: address,
+        /// This table is a list of prizes
+        /// We will have each index correspond to a list of prizes i.e.
+        /// 0: 50000 SUI
+        /// 1: 100000 SUI
+        /// 2: 200000 SUI
+        prizes: vector<Balance<Asset>>,
+        /// This is the prize odds for each of the prizes in the vector
+        prize_odds: vector<u64>,
+        prize_payouts: vector<u64>,
+        public_key: vector<u8>,
+	}
+
+    /// Initializes the store with all of the lottery tickets.
+    /// We allow the user of the store.
+    /// We purposely design the convenience store to be an owned object so that we
+    /// don't need to make this in a shared format.
+    public fun open_store<Asset>(
+        coin: Coin<Asset>, 
+        prizes: vector<u64>,
+        prize_odds: vector<u64>, 
+        prize_payouts: vector<u64>,
+        public_key: vector<u8>,
+        ctx: &mut TxContext
+    ): Coin<Asset> {
+        let prizes_len = vector::length(&prizes);
+        let odds_len = vector::length(&prizes);
+        assert!(prizes_len == odds_len, EInvalidInputs);
+        let coin_balance = coin::into_balance(coin);
```

Comment:
nit: This `balance` ends up getting turned back into a coin at the end -- might as well just modify it directly?